rtape.c

```diff
@@ -0,0 +1,647 @@
+}
+
+static void packet_los(const unsigned char *data, int len)
+{
+  cmd_close("Connection error", 0);
+}
+
+static void packet_cls(const unsigned char *data, int len)
+{
+  cmd_close("Connection closed", 0);
+}
+
+static void packet_dat(const unsigned char *data, int len)
+{
+  state(data, len);
+}
+
+static void state_ignore(const unsigned char *data, int len)
+{
+  (void)data;
+  (void)len;
+}
+
+static void state_version(const unsigned char *data, int len)
+{
+  if (strncasecmp((const char *)data, version, strlen(version)) != 0) {
+    fprintf(log, "Bad record version\n");
+    send_packet(CHOP_CLS, NULL, 0);
+    return;
+  }
+
+  send_packet(CHOP_DAT, version, strlen(version));
+
+  state = state_opcode;
+}
+
+static void state_opcode(const unsigned char *data, int len)
+{
+  command_opcode = data[0];
+  state = state_len1;
+  if (len > 1)
+    state(data + 1, len - 1);
+}
+
+static void state_len1(const unsigned char *data, int len)
+{
+  command_len = (int)data[0] << 8;
+  state = state_len2;
+  if (len > 1)
+    state(data + 1, len - 1);
+}
+
+static void state_len2(const unsigned char *data, int len)
+{
+  command_len |= data[0];
+  state = state_data;
+  command_ptr = command_data;
+  state(data + 1, len - 1);
+}
+
+static void state_data(const unsigned char *data, int len)
+{
+  int n = MIN(len, command_len - (command_ptr - command_data));
+  memcpy(command_ptr, data, n);
+  command_ptr += n;
+  if (command_ptr - command_data == command_len) {
+    dispatch(command_opcode, MAX_COMMANDS, command_handler,
+             command_data, command_len);
+    state = state_opcode;
+    data += n;
+    len -= n;
+    if (len > 0)
+      state(data, len);
+  }
+}
+
+static void send_command(int command, const void *data, size_t len)
+{
+  char buf[MAX_PACKET];
+  const char *p;
+  size_t n;
+
+  buf[0] = command;
+  buf[1] = (len >> 8) & 0xFF;
+  buf[2] = len & 0xFF;
+  n = MIN(len, 450);
+  if (n > 0)
+    memcpy(buf + 3, data, n);
+  len -= n;
+  send_packet(CHOP_DAT, buf, n + 3);
+
+  for (p = data + n; len > 0; p += n, len -= n) {
+    n = MIN(len, 450);
+    send_packet(CHOP_DAT, p, n);
+  }
+}
+
+static void cmd_login(const unsigned char *data, int len)
+{
+  char buf[1];
+  fprintf(log, "Received login: %s\n", data);
+  memset(buf, 0, sizeof buf);
+  send_command(CMD_LGR, buf, sizeof buf);
+}
+
+static char *
+parse(char **p)
+{
+  char *v;
+  v = *p;
+  while (**p != 0 && **p != ' ')
+    (*p)++;
+  **p = 0;
+  (*p)++;
+  return v;
+}
+
+static void cmd_mount(const unsigned char *data, int len)
+{
+  char buf[MAX_PACKET];
+  char *p, *type, *reel, *drive, *size, *density;
+
+  strncpy(buf, (const char *)data, len);
+
+  p = buf;
+  type = parse(&p);
+  reel = parse(&p);
+  drive = parse(&p);
+  size = parse(&p);
+  density = parse(&p);
+
+  while (*p) {
+    char *option = parse(&p);
+    if (strcasecmp(option, "NOREWIND") == 0)
+      flags |= FLG_NOREW;
+  }
+
+  fprintf(log, "Mount: type=%s, reel=%s, drive=%s, size=%s, density=%s\n",
```

Comment:
If it's not too messy, please add who (which host) caused the mount. There may be several concurrent RTAPE sessions, right?

Reply:
The peer is reported when the RFC comes in.  I don't know if there can be concurrent sessions.  I expected the API to provide some kind of session/connection identifier, but I didn't see any.  I initially planned to have a  table for multiple sessions, but I dropped that.  So how do you track sessions?

Reply:
So save the remote peer when the RFC comes in? This program only handles one session and then terminates. To allow more than one session, you probably should do the sequence (1) make socket, (2) LSN, wait for RFC, (3) OPN, and (4a) spawn a thread for that socket while (4b) creating a new socket, doing a new LSN etc. Check the python servers for structure, perhaps? If you only want one session at a time (would be reasonable if you have one real tape drive in mind), do 4a+4b sequentially instead.

Reply:
Fixed.